tools/shared.py

```diff
@@ -1826,7 +1801,35 @@ def link_lld(args, target, opts=[], all_external_symbols=None):
       if not Settings.RELOCATABLE:
         cmd.append('--global-base=%s' % Settings.GLOBAL_BASE)
 
-    cmd += opts
+    return cmd
+
+  @staticmethod
+  def link_lld(args, target, external_symbol_list=None):
+    if not os.path.exists(WASM_LD):
+      exit_with_error('linker binary not found in LLVM directory: %s', WASM_LD)
+    # runs lld to link things.
+    # lld doesn't currently support --start-group/--end-group since the
+    # semantics are more like the windows linker where there is no need for
+    # grouping.
+    args = [a for a in args if a not in ('--start-group', '--end-group')]
+
+    # Emscripten currently expects linkable output (SIDE_MODULE/MAIN_MODULE) to
+    # include all archive contents.
+    if Settings.LINKABLE:
+      args.insert(0, '--whole-archive')
+      args.append('--no-whole-archive')
+
+    if Settings.STRICT:
+      args.append('--fatal-warnings')
+
+    cmd = [WASM_LD, '-o', target] + args
+    for a in Building.llvm_backend_args():
+      cmd += ['-mllvm', a]
+
+    # For relocatable output (generating an object file) we don't pass any of the normal linker
+    # flags that are used when building and exectuable
+    if '--relocatable' not in args and '-r' not in args:
+      cmd += Building.calc_lld_flags(external_symbol_list)
```

Comment:
perhaps the name could be more specific? a bunch of flags are added in this function before this call. what is the key difference between them?

Reply:
Right, maybe I should call this `calc_lld_flags_for_exectuable` or add some more comments?

Reply:
`_for_executable` sgtm